### PR TITLE
Implemented convolutional layers (CNN) to extend the model for image data processing.

### PR DESCRIPTION
**Pull request by DhanashreePetare**
### **This PR implements Convolutional Neural Network (CNN) support by introducing the necessary ConvLayer, Flatten and Dense components into pydeepflow/model.py.**

1. Created an independent cnn_runner.py file for testing this extended work.
2. Configured parameters such as kernel size, strides, and padding shown below.
Screenshot 2025-10-02 232720

3. Tested with an example dataset of digits and got following analysis
Screenshot 2025-10-02 232807

### CODE REVIEW

File: pydeepflow/model.py
Comment (ravin-d-27) [inline pydeepflow/model.py:1]: **Review: Conv Layer Implementation**

Hi! Thanks for contributing the convolutional layer implementation. The `im2col`/`col2im` functions and `ConvLayer` class are well-written and the math looks correct. However, there's a structural issue that prevents this from being usable as-is.

**The Problem:**

The Conv and Flatten layers exist as standalone classes, but they're not integrated into the training pipeline. `Multi_Layer_ANN` only trains dense layers - it doesn't know how to forward/backward through Conv layers. This means:

1. When users try to build CNNs (like in `cnn_runner.py`), they have to manually run forward passes through Conv layers
2. During training, gradients never reach the Conv layers, so their weights never update
3. The Conv filters remain frozen at their random initialization values


**Suggested Fix:**

Instead of modifying `Multi_Layer_ANN` (which should stay focused on dense networks), please create a separate `Multi_Layer_CNN` class that:

1. Accepts a layer configuration with Conv, Activation, Flatten, and Dense layers
2. Implements `forward_propagation()` that chains through all layer types
3. Implements `backpropagation()` that propagates gradients backward through Conv layers
4. Updates weights for both Conv and Dense layers in a single training loop

This keeps the codebase clean - `Multi_Layer_ANN` for dense networks, `Multi_Layer_CNN` for convolutional ones.

**Example Structure:**

```python
class Multi_Layer_CNN:
    def __init__(self, input_shape, conv_layers, dense_layers, ...):
        # Build conv layers: ConvLayer objects
        # Build flatten layer
        # Build dense layers
        
    def forward_propagation(self, X):
        # Forward through conv → flatten → dense
        # Cache activations for backward pass
        
    def backpropagation(self, ...):
        # Backward through dense → flatten → conv
        # Update all weights
```

**Comment by DhanashreePetare** [comment]
Hi @ravin-d-27, 
Thank you for the detailed review! I've implemented all the requested changes:

### **Key Changes**
1. ✅ Created Separate [Multi_Layer_CNN] Class
- Built dedicated CNN class without modifying existing [Multi_Layer_ANN](keeps dense-only focus)
- Accepts flexible layer configuration: {'type': 'conv'/'flatten'/'dense', ...}
- Maintains clean separation: ANN for dense networks, CNN for convolutional networks


2. ✅ Implemented End-to-End Training Pipeline
- Forward Propagation: Chains Conv → Flatten → Dense layers sequentially
- Backpropagation: Propagates gradients backward through all layer types (Dense → Flatten → Conv)
- Weight Updates: Updates both Conv and Dense layer weights in single training loop


3. ✅ Unified Parameter Management
- All trainable parameters (Conv W/b + Dense W/b) stored in [self.trainable_params]
- Single optimizer call updates all weights: [optimizer.update(self.trainable_params, gradients)]
- Proper gradient flow ensures Conv filters learn during training (no longer frozen)

4. ✅ Working Demonstration
- Updated [cnn_runner.py] with proper layer configuration
- Successfully trains Conv → Conv → Flatten → Dense → Dense architecture
- Achieves 95.28% accuracy on digits classification (proves Conv layers are learning)

### **Result**
✅ Fully functional CNN training pipeline where Conv layer weights properly update during training, solving the core issue identified in the review.

The implementation now successfully trains CNN models with proper weight updates for both Conv and Dense layers, achieving 95.28% accuracy on the digits dataset.

Please review the changes and let me know if something still need correction. Thanks!

**Review by ravin-d-27 — APPROVED** [review]
Hi @DhanashreePetare ! Looks good to me. Good work. I am Merging this PR. Thank you!
